### PR TITLE
Zec3

### PR DESCRIPTION
**Pull request by makoto**
https://github.com/ensdomains/address-encoder/pull/10 seemed a bit out of date so I cherry picked some of commits and updated to use the latest `bs58Encode/bs58Decode`

### CODE REVIEW

**Comment by makoto** [comment]
Umm, I also plugged in a few test from Trust Wallet https://github.com/trustwallet/ens-coincodec/blob/master/zcash_test.go#L14 and tests are failing. Either this PR or Trust Wallet implementation must be incorrect

**Comment by makoto** [comment]
Looks like Trust Wallet version was simply doing base58. Contacted them and they are taking a look

**Comment by makoto** [comment]
Failing test should pass once merged with https://github.com/ensdomains/address-encoder/pull/46

File: src/index.ts [hunk 1]
```diff
@@ -21,38 +24,54 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
         ) {
           throw Error('Unrecognised address format');
         }
-        addr = Buffer.concat([Buffer.from([p2pkhVersion]), data.slice(3, 3 + data.readUInt8(2))]);
+        if (p2pkhVersion[0] instanceof Array) {
```
Comment (Arachnid) [inline src/index.ts:27]: Why do we have an array if we ignore all elements but the first?

File: src/index.ts [hunk 2]
```diff
@@ -21,38 +24,54 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
...
+        addr = Buffer.concat([Buffer.from([p2pkhVersion[0]]), data.slice(3, 3 + data.readUInt8(2))]);
         // @ts-ignore
         return bs58Encode(addr);
       case 0xa9: // P2SH: OP_HASH160 <scriptHash> OP_EQUAL
         if (data.readUInt8(data.length - 1) !== 0x87) {
           throw Error('Unrecognised address format');
         }
-        addr = Buffer.concat([Buffer.from([p2shVersion]), data.slice(2, 2 + data.readUInt8(1))]);
+        if (p2shVersion[0] instanceof Array) {
+          addr = Buffer.concat([Buffer.from(p2shVersion[0].slice(0, 2)), data.slice(2, 2 + data.readUInt8(1))]);
```
Comment (Arachnid) [inline src/index.ts:39]: Why slice the array, rather than require it be a certain length?

File: src/index.ts [hunk 3]
```diff
@@ -21,38 +24,54 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
...
+  p2shVersions: (number | number[])[],
+): (data: string) => Buffer {
   return (data: string) => {
     const addr = bs58Decode(data);
-    const version = addr.readUInt8(0);
-    if (p2pkhVersions.includes(version)) {
-      return Buffer.concat([Buffer.from([0x76, 0xa9, 0x14]), addr.slice(1), Buffer.from([0x88, 0xac])]);
-    } else if (p2shVersions.includes(version)) {
-      return Buffer.concat([Buffer.from([0xa9, 0x14]), addr.slice(1), Buffer.from([0x87])]);
+    const version = addr.filter((b:any) => addr.slice(-20).indexOf(b) === -1);
```
Comment (Arachnid) [inline src/index.ts:56]: Can you document what's going on here in a comment? I can't follow it.

File: src/index.ts [hunk 4]
```diff
@@ -21,38 +24,54 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
...
+    if (p2pkhVersions.some(b => (b instanceof Array) ? version.every((v:any, i:any) => v === b[i]) : version.includes(b))) {
```
Comment (Arachnid) [inline src/index.ts:57]: Likewise here. Examples of how this function can be called would be useful.

**Review by Arachnid — CHANGES_REQUESTED** [review]
Why is the build currently failing? Can the bugfix be moved into this PR so we don't have to merge a failing PR?

**Comment by Arachnid** [comment]
I don't see any provision for shielded addresses here. Are they supported? Do we have a path forward to adding them if not?

File: src/index.ts (hunk 1 above)
Comment (makoto) [reply src/index.ts:30]: I didn't write this part but I am assuming that the original contributor responded to your comment where you asked him to refactor to take `either a buffer or an array of bytes, and check for the version inside that array`, hence it's checking which pattern it is.

https://github.com/ensdomains/address-encoder/pull/10#discussion_r349373290

File: src/index.ts (hunk 2 above)
Comment (makoto) [reply src/index.ts:42]: Again, this part is not done by me. slice seems okay to me. What else do you suggest when you say (`require it be a certain length`). Just to check `p2shVersion[0] === 2`?

**Comment by makoto** [comment]
@Arachnid I don't know how shielded addresses work and asked for support at Zcash discord channel if they have some reference implementation on their library (ideally js) but go no response.

File: src/index.ts (hunk 3 above)
Comment (makoto) [reply src/index.ts:59]: Good point, I have no idea why `-20`. I only fixed the syntax error.

File: src/index.ts (hunk 4 above)
Comment (makoto) [reply src/index.ts:60]: I didn't write majority of code so hard to document the intention. If we don't get feedback from the original contributor, we could just close this as I don't fully understand either.

**Comment by Arachnid** [comment]
I think we need to close this and wait for someone to submit a version that's tidy and we can be sure is functioning as intended.

**Pull request closed without merging**